p4runtime_sh/shell.py

```diff
@@ -2410,15 +2411,32 @@ def sniff(self, function=None, timeout=None):
         the function to every packet-in message.
         """
         msgs = []
-        while True:
-            try:
-                msgs.append(self.packet_in_queue.get(block=True, timeout=timeout))
-            except queue.Empty:
-                # No item available when timeout.
-                break
-            except KeyboardInterrupt:
-                # User sends an interrupt(e.g., Ctrl+C).
-                break
+
+        if timeout is not None and timeout < 0:
+            raise ValueError("Timeout can\'t be a negative number.")
+
+        if timeout is None:
+            while True:
+                try:
+                    msgs.append(self.packet_in_queue.get(block=True))
+                except KeyboardInterrupt:
+                    # User sends a Ctrl+C -> breaking
+                    break
+
+        else:  # timeout parameter is provided
+            remaining_time = timeout
+            while remaining_time > 0:
+                try:
+                    start = time.time()
+                    msgs.append(self.packet_in_queue.get(block=True, timeout=remaining_time))
+                    elapsed = time.time() - start
+                    remaining_time -= elapsed
```

Comment:
I think it would be better (more accurate?) to have something like this:
```python
deadline = time.time() + timeout
remaining_time = timeout
while remaining_time > 0:
    try:
        msgs.append(self.packet_in_queue.get(block=True, timeout=remaining_time))
        remaining_time = deadline - time.time()
```

Reply:
Sure, I think it's also more intuitive.